### PR TITLE
Add 3rd derivatives

### PR DESCRIPTION
**Pull request by andysim**
These are needed for GGA hessians.

The code compiles as-is, but I'll need some help with the testing.  I bumped up the derivative level in the existing deriv tests, and `python setup.py test` gave the following:
```
platform darwin -- Python 3.7.1, pytest-4.0.2, py-1.7.0, pluggy-0.8.0
rootdir: /Users/andysim/programming/gau2grid, inifile:
plugins: remotedata-0.3.1, openfiles-0.3.1, doctestplus-0.2.0, arraydiff-0.3
collected 140 items                                                                                                           

gau2grid/tests/test_c_code.py sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 63%]
ssssssssssssssss                                                                                                        [ 75%]
gau2grid/tests/test_c_generator.py ..................                                                                   [ 87%]
gau2grid/tests/test_order.py ..........                                                                                 [ 95%]
gau2grid/tests/test_rsh.py .......                                                                                      [100%]

=========================================== 35 passed, 105 skipped in 4.18 seconds ============================================
```

Any suggestions for testing are very much welcomed.

### CODE REVIEW

**Comment by dgasmith** [comment]
This is a bit awkward to test I do something like:

```
cmake -H. -Bobjdir \
    -DMAX_AM=6 \
    -DCMAKE_C_COMPILER=clang \
    -DCMAKE_BUILD_TYPE=debug \
    -DPYTHON_EXECUTABLE=...
cd objdir
make VERBOSE=1
```

and then symlink the built SO into the `gau2grid` dev folder. In theory we can use pip to manage this, but I have yet to figure that one out.

**Comment by dgasmith** [comment]
This pull request **introduces 1 alert** when merging f708b033e343424246ffd2b13b2e64a4d245e9b2 into 562ed3ccf7fdc76c5d309befef6075e2fc6613bf - [view on LGTM.com](https://lgtm.com/projects/g/dgasmith/gau2grid/rev/pr-576c9eadfc116c1439bfdc975f5101b99be4209c)

**new alerts:**

* 1 for Unused local variable

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: gau2grid/c_wrapper.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 # First check the local folder
 try:
     abs_path = os.path.dirname(os.path.abspath(__file__))
-    cgg = np.ctypeslib.load_library("gg", abs_path)
+    cgg = np.ctypeslib.load_library("libgg", abs_path)
```
Comment (andysim) [inline gau2grid/c_wrapper.py]: Question for you, @loriab.  I had to make this change here for macOS (simply making a symlink called `gg` didn't work).  Do you know if this is a good idea?  No problem if not - I can continue down the ctypes documentation rabbithole I dived down yesterday.

Comment (andysim) [reply gau2grid/c_wrapper.py]: Well, I think I have the answer - Travis didn't like the change.  Will revert it and figure out a different strategy for macOS....

Comment (loriab) [reply gau2grid/c_wrapper.py]: I'm not sure what's going on with `load_library`. DGAS [just did the opposite](https://github.com/dgasmith/gau2grid/commit/bdd46992654239f5eb08b9db72052779ebf6d17c) of what you propose a few weeks ago. And according to the docs, either should work. Plus, I can't track why it needed to change in the July case. I know in the c-f build, Mac likes the current master.

Sorry, that's all pretty useless. One point of info I can add is that the linux lib with SONAME set is renamed when it's copied to the python pkg directory and that happens https://github.com/dgasmith/gau2grid/blob/master/CMakeLists.txt#L159.

**Comment by dgasmith** [comment]
This pull request **introduces 1 alert** when merging 5b81b8ba681c6243e8a01796792b0c8919610e69 into 562ed3ccf7fdc76c5d309befef6075e2fc6613bf - [view on LGTM.com](https://lgtm.com/projects/g/dgasmith/gau2grid/rev/pr-dad2a568162bf8fa2e55cb866f2250721841458e)

**new alerts:**

* 1 for Unused local variable

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: gau2grid/c_wrapper.py (hunk 1 above)
Comment (andysim) [reply gau2grid/c_wrapper.py]: Thanks for that info - very helpful.  I have no clue what I'm doing, but reverting the change (to match current master) fixes the travis build.  If I do that on my laptop (running Mojave), the library fails to load and most of the pytests are skipped.  Very strange, but at least I know a workaround for local development - just changing the `gg` to `libgg` does the trick.

Comment (dgasmith) [reply gau2grid/c_wrapper.py]: Maybe something stupid?

```python
cgg = np.ctypeslib.load_library("gg", abs_path)
if cgg is None:
    cgg = np.ctypeslib.load_library("libgg", abs_path)
```

Comment (loriab) [reply gau2grid/c_wrapper.py]: maybe make a note of your current numpy version, then update numpy and see if that's happier with the plain "gg"

Comment (andysim) [reply gau2grid/c_wrapper.py]: I suspect my numpy version 1.16 could be at fault, but I hate the idea that the code is version dependent.  I implemented something similar to DGAS's suggestion - the function throws if it doesn't find the right library, so I just threw in a nested `try/except` block.

**Comment by dgasmith** [comment]
This pull request **introduces 1 alert** when merging 9d6a8dd5f4aea296cef4b2630562d56b9c58b915 into 562ed3ccf7fdc76c5d309befef6075e2fc6613bf - [view on LGTM.com](https://lgtm.com/projects/g/dgasmith/gau2grid/rev/pr-25e04b068d9ae7192d2684ab9cbd368b80e58dc9)

**new alerts:**

* 1 for Unused local variable

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: gau2grid/c_generator.py [hunk 2]
```diff
@@ -307,11 +307,11 @@ def shell_c_generator(cg, L, function_name="", grad=0, cartesian_order="row", in
...
     paritioned_loops = False
-    if (grad == 1) and (L >= 3):
+    if (grad == 1) and (L >= 7):
         paritioned_loops = True
-    elif (grad == 2) and (L >= 1):
+    elif (grad == 2) and (L >= 3):
         paritioned_loops = True
-    elif (grad == 3):
+    elif (grad == 3) and (L >= 2):
         paritioned_loops = True
```
Comment (andysim) [inline gau2grid/c_generator.py]: While messing with the value for this, I found that I could bump up the AM for the 1st and 2nd derivs without violating the L1 cache metric.  If this is not correct, please let me know and I'll revert it.

File: gau2grid/c_wrapper.py [hunk 3]
```diff
@@ -23,7 +23,12 @@
     __libgg_path = os.path.join(abs_path, cgg._name)
     __lib_found = True
 except OSError:
-    cgg = None
+    try:
+        cgg = np.ctypeslib.load_library("libgg", abs_path)
```
[the remark this answers is not in the document]
Comment (andysim) [reply gau2grid/c_wrapper.py]: `gg` is tried first, on line 22, and then `libgg` as a fallback here

**Comment by andysim** [comment]
I'll add some docs later today.  This one *should* be good to go.  The only testing thus far is checking that the reference code and the C version are consistent.  When I get to hooking it into the GGA hessian code it will be more thoroughly stress tested.

**Pull request merged**